### ISSUE TITLE
[HOLD for payment] [$250] BUG: Prevent editing messages sent to Chronos

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#11689]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Send any message to chronos
2. Edit message sent to chronos
3. See Auth UpdateReportComment returned an error
3. Notice that the edits were not recognized, either

## Expected Result:
User should not be to initiate editing messages sent to Chronos

## Actual Result:
User is able to try to edit their message. An error message `Auth UpdateReportComment` is surfaced and the edit wasn’t recognized. A red dot is also shown besides chronos in the LHN.

## Workaround:
Sign out and sign back in to clear the error red dot.

## Platform:

Where is this issue occurring?

- All platforms

**Version Number:** 1.2.12-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/194736175-6f4bcf82-5fcb-4350-8d28-e892a7e235ca.mp4

<img width="561" alt="image (1)" src="https://user-images.githubusercontent.com/43996225/194736185-824e07ae-92e6-43f6-8f22-13587835b654.png">

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @michaelhaxhiu 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1665171071433639

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by michaelhaxhiu** [comment]
I'll take it :)

**Comment by abekkala** [comment]
Oh! thanks @michaelhaxhiu  I was just about to apply the label for assignment. thanks for taking!

**Comment by chiragsalian** [comment]
Shall investigate today

**Comment by NikkiWines** [comment]
Sorry haven't had a chance to look at this issue this week but I'll take a look tomorrow or early next week

**Comment by NikkiWines** [comment]
Quick question - why do we want to allow edits to Chronos messages? 

It's not really functionality we've offered historically and we specifically prevent edits for Chronos chats [here](https://github.com/Expensify/Auth/blob/10fbc5ad2522d7801a61afbea4a4e0c0c76bb146/auth/command/UpdateReportComment.cpp#L37-L39). Since it's been used for historical reference purposes in the past, I'm not sure we want to allow editing of chronos comments.

**Comment by michaelhaxhiu** [comment]
@NikkiWines that's a great philosophical question. I agree with your logic and reasoning. I bet we can just coin a rule here... anytime you talk to a bot, you can't change your messages.

**Comment by michaelhaxhiu** [comment]
If we arrive at this conclusion, then let's block the ability to edit these messages altogether.

**Comment by NikkiWines** [comment]
Not sure where to discuss this further but yeah I think let's either prevent editing (which I'm not sure is necessary since we already do that just in a more inelegant way) or just #donothing and leave it as it

**Comment by michaelhaxhiu** [comment]
Was thinking a quick post in #expensify-open-source? If we #donothing then this bug will persist anytime someone tries to edit messages to chronos. I feel like we should prevent editing the messages there for that reason.

**Comment by NikkiWines** [comment]
Sounds good, thanks for posting therre @michaelhaxhiu!

**Comment by michaelhaxhiu** [comment]
Based on slack discussion let's mark this as `external` and it should assign a BZ to post to upwork. 

We think the job value should be capped at $500 since this is straightforward/easy (but start price at $250). That's the gist from slack post. 🚀

**Comment by NikkiWines** [comment]
Updated the issue and title appropriately, and going to assign the `external` label so we can get someone working on this 🚀

**Comment by jayeshmangwani** [comment]
# Proposal

1. we have to create a new function that can check if the report object has any participants with CHRONOS mail in [src/libs/ReportUtils.js file](https://github.com/Expensify/App/blob/main/src/libs/ReportUtils.js) file

```diff
+/**
+ * Returns true if Chronos is one of the chat participants (1:1)
+ * @param {Object} report
+ * @returns {Boolean}
+ */
+function chatIncludesChronos(report) {
+    return report.participants
+            && _.contains(report.participants, CONST.EMAIL.CHRONOS);
+}
```

2. we have to pass a new param with like this 

```diff
<MiniReportActionContextMenu
    reportID={this.props.report.reportID}
    reportAction={this.props.action}
    isArchivedRoom={ReportUtils.isArchivedRoom(this.props.report)}
    displayAsGroup={this.props.displayAsGroup}
    isVisible={
        hovered
        && !this.state.isContextMenuActive
        && !this.props.draftMessage
    }
    draftMessage={this.props.draftMessage}
+  isChronosRoom={ReportUtils.chatIncludesChronos(this.props.report)}
/>
```

3. we have to pass new param to the contextAction.shouldShow function

https://github.com/Expensify/App/blob/e434efa1c455620c510441628b94081e0f8a564e/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.js#L40

```diff
-const shouldShowFilter = contextAction => contextAction.shouldShow(this.props.type, this.props.reportAction, this.props.isArchivedRoom, this.props.betas, this.props.anchor);
+const shouldShowFilter = contextAction => contextAction.shouldShow(this.props.type, this.props.reportAction, this.props.isArchivedRoom, this.props.betas, this.props.anchor, this.props.isChronosRoom);
```

4. we have to add an one more condition that if chat is not the chronos then shouldShow will be true

https://github.com/Expensify/App/blob/e434efa1c455620c510441628b94081e0f8a564e/src/pages/home/report/ContextMenu/ContextMenuActions.js#L165-L167

```diff
-shouldShow: (type, reportAction, isArchivedRoom) => (
-    type === CONTEXT_MENU_TYPES.REPORT_ACTION && ReportUtils.canEditReportAction(reportAction) && !isArchivedRoom
-),
+shouldShow: (type, reportAction, isArchivedRoom, betas, menuTarget, isChronosRoom) => (
+    type === CONTEXT_MENU_TYPES.REPORT_ACTION && ReportUtils.canEditReportAction(reportAction) && !isArchivedRoom && !isChronosRoom
+),
```

## video after fix
https://user-images.githubusercontent.com/35371050/199975543-466a70c0-c55b-4193-b11e-1d8b3aa26363.mov

**Comment by sobitneupane** [comment]
@roryabraham Since we are preventing users to edit message, do we want to prevent deleting message as well?

**Comment by sobitneupane** [comment]
@jayeshmangwani Your proposal looks good to me. Few suggestions:

- Use `isChronosReport` instead of `isChronosRoom`
- Make sure to pass `isChronosReport` to `<PopoverReportActionContextMenu />` as well. It is used in native devices and mWeb.
https://github.com/Expensify/App/blob/e434efa1c455620c510441628b94081e0f8a564e/src/pages/home/report/ReportActionsView.js#L376

@jayeshmangwani's [Proposal](https://github.com/Expensify/App/issues/11689#issuecomment-1303465185) looks good to me.

🎀👀🎀 C+ reviewed

cc: @roryabraham

**Comment by tjferriss** [comment]
Upworks job post is live: https://www.upwork.com/ab/applicants/1588657705516900352/job-details

@jayeshmangwani and @sobitneupane can you both please apply?

**Comment by sobitneupane** [comment]
@tjferriss The [proposal](https://github.com/Expensify/App/issues/11689#issuecomment-1303465185) is not accepted yet. Waiting approval from @roryabraham.

**Comment by roryabraham** [comment]
Whoa, super sorry I missed this everyone.

> Since we are preventing users to edit message, do we want to prevent deleting message as well?

Yes, prevent delete as well as edit.

**Comment by jayeshmangwani** [comment]
PR is up
cc: @sobitneupane @roryabraham

**Comment by roryabraham** [comment]
PR merged

**Comment by roryabraham** [comment]
Created a follow-up PR [here](https://github.com/Expensify/App/pull/12875) to clean up some propTypes inconsistencies/mistakes that I saw

**Comment by roryabraham** [comment]
Just waiting on deploy, QA, and payment

**Comment by roryabraham** [comment]
This was deployed to production 5 days ago. cc @tjferriss

**Comment by tjferriss** [comment]
Ok I'm manually adding the checklist to close this out:

- [ ] If this bug was caught on production, add or update the [regression tests](https://expensify.testrail.io/index.php?/dashboard) ([SO for How do I get new regular regression QA tests added?](https://stackoverflow.com/c/expensify/questions/363/364#364)) so that the same bug can’t reach production again and link to the updated test here:

- [ ] Link to the PR that caused the bug:

- [ ] Comment on the offending PR to notify the author and reviewers that the PR caused this bug.

- [ ] Reduce the pay of the C+ on the offending PR if they should have caught the regression (ie. they overlooked a step/checkbox that led to the regression.)

- [ ] Start a discussion in #contributor-plus about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

- [ ] Complete payment to @contrib1, @cplus1, and @issureporter1 [after 7 days](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#payment-for-contributions).

**Comment by tjferriss** [comment]
Thanks for the reminder. It looks like the job post expired. I just set up a new one here - https://www.upwork.com/jobs/~0162b815998a19498c.

@jayeshmangwani and @sobitneupane can you both please apply? Let me know when you do here and I'll process payment promptly.

**Comment by jayeshmangwani** [comment]
@tjferriss Applied for the job

**Comment by sobitneupane** [comment]
@tjferriss Applied. This is eligible for 50% bonus as well.

**Comment by tjferriss** [comment]
Ok payments have been sent.

**Comment by tjferriss** [comment]
@jayeshmangwani / @roryabraham can you please complete the steps [here](https://github.com/Expensify/App/issues/11689#issuecomment-1331072208).

**Comment by roryabraham** [comment]
done

**Comment by tjferriss** [comment]
Ok it looks like we've checked off all the tasks so I'm going to close this out.
